diff --git a/Microsoft/UI/Xaml/Markup/Compiler/CodeGen/ApiInformationCodeGenerator.cs b/Microsoft/UI/Xaml/Markup/Compiler/CodeGen/ApiInformationCodeGenerator.cs
index 64777b4..2e70920 100644
--- a/Microsoft/UI/Xaml/Markup/Compiler/CodeGen/ApiInformationCodeGenerator.cs
+++ b/Microsoft/UI/Xaml/Markup/Compiler/CodeGen/ApiInformationCodeGenerator.cs
@@ -9,8 +9,12 @@ internal class ApiInformationCodeGenerator : CodeGeneratorBase<ApiInformation>,
 	{
 		get
 		{
-			ICodeGenOutput callExpression = base.Instance.Method.CodeGen().CallExpression;
-			return new LanguageSpecificString(() => string.Format("{0}({1})", callExpression.CppCXName(), string.Join(", ", base.Instance.Parameters.Select((ApiInformationParameter p) => p.CodeGen().CallExpression.CppCXName()))), () => string.Format("{0}({1})", callExpression.CppWinRTName(), string.Join(", ", base.Instance.Parameters.Select((ApiInformationParameter p) => p.CodeGen().CallExpression.CppWinRTName()))), () => string.Format("{0}({1})", callExpression.CSharpName(), string.Join(", ", base.Instance.Parameters.Select((ApiInformationParameter p) => p.CodeGen().CallExpression.CSharpName()))), () => string.Format("{0}({1})", callExpression.VBName(), string.Join(", ", base.Instance.Parameters.Select((ApiInformationParameter p) => p.CodeGen().CallExpression.VBName()))));
+			ICodeGenOutput callExpression = Instance.Method.CodeGen().CallExpression;
+			return new LanguageSpecificString(
+				() => string.Format("{0}({1})", callExpression.CppCXName(), string.Join(", ", Instance.Parameters.Select((ApiInformationParameter p) => p.CodeGen().CallExpression.CppCXName()))),
+				() => string.Format("{0}({1})", callExpression.CppWinRTName(), string.Join(", ", Instance.Parameters.Select((ApiInformationParameter p) => p.CodeGen().CallExpression.CppWinRTName()))),
+				() => string.Format("{0}({1})", callExpression.CSharpName(), string.Join(", ", Instance.Parameters.Select((ApiInformationParameter p) => p.CodeGen().CallExpression.CSharpName()))),
+				() => string.Format("{0}({1})", callExpression.VBName(), string.Join(", ", Instance.Parameters.Select((ApiInformationParameter p) => p.CodeGen().CallExpression.VBName()))));
 		}
 	}
 }
