diff --git a/Microsoft.Windows.UI.Xaml.Build.Tasks.csproj b/Microsoft.Windows.UI.Xaml.Build.Tasks.csproj
index 572f39c..c9e3111 100644
--- a/Microsoft.Windows.UI.Xaml.Build.Tasks.csproj
+++ b/Microsoft.Windows.UI.Xaml.Build.Tasks.csproj
@@ -3,6 +3,8 @@
     <AssemblyName>Microsoft.Windows.UI.Xaml.Build.Tasks</AssemblyName>
     <GenerateAssemblyInfo>False</GenerateAssemblyInfo>
     <TargetFramework>net47</TargetFramework>
+    <TransformOnBuild>true</TransformOnBuild>
+    <TransformOutOfDateOnly>false</TransformOutOfDateOnly>
   </PropertyGroup>
   <PropertyGroup>
     <LangVersion>12.0</LangVersion>
@@ -12,6 +14,162 @@
     <RootNamespace />
   </PropertyGroup>
   <ItemGroup />
+  <ItemGroup>
+    <Compile Update="Microsoft\Xaml\XamlCompiler\CodeGenerators\WRL_BindingInfoPass1.cs">
+      <AutoGen>True</AutoGen>
+      <DesignTime>True</DesignTime>
+      <DependentUpon>WRL_BindingInfoPass1.tt</DependentUpon>
+    </Compile>
+    <Compile Update="Microsoft\Xaml\XamlCompiler\CodeGenerators\WRL_BindingInfoPass2.cs">
+      <AutoGen>True</AutoGen>
+      <DesignTime>True</DesignTime>
+      <DependentUpon>WRL_BindingInfoPass2.tt</DependentUpon>
+    </Compile>
+    <Compile Update="Microsoft\Xaml\XamlCompiler\CodeGenerators\WRL_PageBinding.cs">
+      <AutoGen>True</AutoGen>
+      <DesignTime>True</DesignTime>
+      <DependentUpon>WRL_PageBinding.tt</DependentUpon>
+    </Compile>
+    <Compile Update="Microsoft\Xaml\XamlCompiler\CodeGenerators\WRL_PageBindingTracking.cs">
+      <AutoGen>True</AutoGen>
+      <DesignTime>True</DesignTime>
+      <DependentUpon>WRL_PageBindingTracking.tt</DependentUpon>
+    </Compile>
+    <Compile Update="Microsoft\Xaml\XamlCompiler\CodeGenerators\WRL_PagePass1.cs">
+      <AutoGen>True</AutoGen>
+      <DesignTime>True</DesignTime>
+      <DependentUpon>WRL_PagePass1.tt</DependentUpon>
+    </Compile>
+    <Compile Update="Microsoft\Xaml\XamlCompiler\CodeGenerators\WRL_PagePass2.cs">
+      <AutoGen>True</AutoGen>
+      <DesignTime>True</DesignTime>
+      <DependentUpon>WRL_PagePass2.tt</DependentUpon>
+    </Compile>
+    <Compile Update="Microsoft\Xaml\XamlCompiler\CodeGenerators\WRL_TemplatedMetadataDelegates.cs">
+      <AutoGen>True</AutoGen>
+      <DesignTime>True</DesignTime>
+      <DependentUpon>WRL_TemplatedMetadataDelegates.tt</DependentUpon>
+    </Compile>
+    <Compile Update="Microsoft\Xaml\XamlCompiler\CodeGenerators\WRL_TypeInfoPass1.cs">
+      <AutoGen>True</AutoGen>
+      <DesignTime>True</DesignTime>
+      <DependentUpon>WRL_TypeInfoPass1.tt</DependentUpon>
+    </Compile>
+    <Compile Update="Microsoft\Xaml\XamlCompiler\CodeGenerators\WRL_TypeInfoPass1Impl.cs">
+      <AutoGen>True</AutoGen>
+      <DesignTime>True</DesignTime>
+      <DependentUpon>WRL_TypeInfoPass1Impl.tt</DependentUpon>
+    </Compile>
+    <Compile Update="Microsoft\Xaml\XamlCompiler\CodeGenerators\WRL_TypeInfoPass2.cs">
+      <AutoGen>True</AutoGen>
+      <DesignTime>True</DesignTime>
+      <DependentUpon>WRL_TypeInfoPass2.tt</DependentUpon>
+    </Compile>
+    <Compile Update="Microsoft\Xaml\XamlCompiler\CodeGenerators\WRL_TypeInfoPass2.cs">
+      <AutoGen>True</AutoGen>
+      <DesignTime>True</DesignTime>
+      <DependentUpon>WRL_TypeInfoPass2.tt</DependentUpon>
+    </Compile>
+    <Compile Update="Microsoft\Xaml\XamlCompiler\CodeGenerators\WRL_XamlMetaDataProviderPass1.cs">
+      <AutoGen>True</AutoGen>
+      <DesignTime>True</DesignTime>
+      <DependentUpon>WRL_XamlMetaDataProviderPass1.tt</DependentUpon>
+    </Compile>
+    <Compile Update="Microsoft\Xaml\XamlCompiler\CodeGenerators\WRL_XamlMetaDataProviderPass2.cs">
+      <AutoGen>True</AutoGen>
+      <DesignTime>True</DesignTime>
+      <DependentUpon>WRL_XamlMetaDataProviderPass2.tt</DependentUpon>
+    </Compile>
+    <Compile Update="Microsoft\Xaml\XamlCompiler\CodeGenerators\WRL_XamlMetaDataProviderPass2.cs">
+      <AutoGen>True</AutoGen>
+      <DesignTime>True</DesignTime>
+      <DependentUpon>WRL_XamlMetaDataProviderPass2.tt</DependentUpon>
+    </Compile>
+    <Compile Update="Microsoft\Xaml\XamlCompiler\CodeGenerators\WRL_XamlMetaDataProviderPass1.cs">
+      <AutoGen>True</AutoGen>
+      <DesignTime>True</DesignTime>
+      <DependentUpon>WRL_XamlMetaDataProviderPass1.tt</DependentUpon>
+    </Compile>
+  </ItemGroup>
+  <ItemGroup>
+    <None Include="Microsoft\Xaml\XamlCompiler\CodeGenerators\WRL_BindingInfoPass1.tt">
+      <Generator>TextTemplatingFilePreprocessor</Generator>
+      <ClassNamespace>Microsoft.Xaml.XamlCompiler.CodeGenerators</ClassNamespace>
+      <CustomToolNamespace>Microsoft.Xaml.XamlCompiler.CodeGenerators</CustomToolNamespace>
+      <LastGenOutput>WRL_BindingInfoPass1.cs</LastGenOutput>
+    </None>
+    <None Include="Microsoft\Xaml\XamlCompiler\CodeGenerators\WRL_BindingInfoPass2.tt">
+      <Generator>TextTemplatingFilePreprocessor</Generator>
+      <ClassNamespace>Microsoft.Xaml.XamlCompiler.CodeGenerators</ClassNamespace>
+      <CustomToolNamespace>Microsoft.Xaml.XamlCompiler.CodeGenerators</CustomToolNamespace>
+      <LastGenOutput>WRL_BindingInfoPass2.cs</LastGenOutput>
+    </None>
+    <None Include="Microsoft\Xaml\XamlCompiler\CodeGenerators\WRL_PageBinding.tt">
+      <Generator>TextTemplatingFilePreprocessor</Generator>
+      <ClassNamespace>Microsoft.Xaml.XamlCompiler.CodeGenerators</ClassNamespace>
+      <CustomToolNamespace>Microsoft.Xaml.XamlCompiler.CodeGenerators</CustomToolNamespace>
+      <LastGenOutput>WRL_PageBinding.cs</LastGenOutput>
+    </None>
+    <None Include="Microsoft\Xaml\XamlCompiler\CodeGenerators\WRL_PageBindingTracking.tt">
+      <Generator>TextTemplatingFilePreprocessor</Generator>
+      <ClassNamespace>Microsoft.Xaml.XamlCompiler.CodeGenerators</ClassNamespace>
+      <CustomToolNamespace>Microsoft.Xaml.XamlCompiler.CodeGenerators</CustomToolNamespace>
+      <LastGenOutput>WRL_PageBindingTracking.cs</LastGenOutput>
+    </None>
+    <None Include="Microsoft\Xaml\XamlCompiler\CodeGenerators\WRL_PagePass1.tt">
+      <Generator>TextTemplatingFilePreprocessor</Generator>
+      <ClassNamespace>Microsoft.Xaml.XamlCompiler.CodeGenerators</ClassNamespace>
+      <CustomToolNamespace>Microsoft.Xaml.XamlCompiler.CodeGenerators</CustomToolNamespace>
+      <LastGenOutput>WRL_PagePass1.cs</LastGenOutput>
+    </None>
+    <None Include="Microsoft\Xaml\XamlCompiler\CodeGenerators\WRL_PagePass2.tt">
+      <Generator>TextTemplatingFilePreprocessor</Generator>
+      <ClassNamespace>Microsoft.Xaml.XamlCompiler.CodeGenerators</ClassNamespace>
+      <CustomToolNamespace>Microsoft.Xaml.XamlCompiler.CodeGenerators</CustomToolNamespace>
+      <LastGenOutput>WRL_PagePass2.cs</LastGenOutput>
+    </None>
+    <None Include="Microsoft\Xaml\XamlCompiler\CodeGenerators\WRL_TemplatedMetadataDelegates.tt">
+      <Generator>TextTemplatingFilePreprocessor</Generator>
+      <ClassNamespace>Microsoft.Xaml.XamlCompiler.CodeGenerators</ClassNamespace>
+      <CustomToolNamespace>Microsoft.Xaml.XamlCompiler.CodeGenerators</CustomToolNamespace>
+      <LastGenOutput>WRL_TemplatedMetadataDelegates.cs</LastGenOutput>
+    </None>
+    <None Include="Microsoft\Xaml\XamlCompiler\CodeGenerators\WRL_TypeInfoPass1.tt">
+      <Generator>TextTemplatingFilePreprocessor</Generator>
+      <ClassNamespace>Microsoft.Xaml.XamlCompiler.CodeGenerators</ClassNamespace>
+      <CustomToolNamespace>Microsoft.Xaml.XamlCompiler.CodeGenerators</CustomToolNamespace>
+      <LastGenOutput>WRL_TypeInfoPass1.cs</LastGenOutput>
+    </None>
+    <None Include="Microsoft\Xaml\XamlCompiler\CodeGenerators\WRL_TypeInfoPass1Impl.tt">
+      <Generator>TextTemplatingFilePreprocessor</Generator>
+      <ClassNamespace>Microsoft.Xaml.XamlCompiler.CodeGenerators</ClassNamespace>
+      <CustomToolNamespace>Microsoft.Xaml.XamlCompiler.CodeGenerators</CustomToolNamespace>
+      <LastGenOutput>WRL_TypeInfoPass1Impl.cs</LastGenOutput>
+    </None>
+    <None Include="Microsoft\Xaml\XamlCompiler\CodeGenerators\WRL_TypeInfoPass2.tt">
+      <Generator>TextTemplatingFilePreprocessor</Generator>
+      <ClassNamespace>Microsoft.Xaml.XamlCompiler.CodeGenerators</ClassNamespace>
+      <CustomToolNamespace>Microsoft.Xaml.XamlCompiler.CodeGenerators</CustomToolNamespace>
+      <LastGenOutput>WRL_TypeInfoPass2.cs</LastGenOutput>
+    </None>
+    <None Include="Microsoft\Xaml\XamlCompiler\CodeGenerators\WRL_XamlMetaDataProviderPass1.tt">
+      <Generator>TextTemplatingFilePreprocessor</Generator>
+      <ClassNamespace>Microsoft.Xaml.XamlCompiler.CodeGenerators</ClassNamespace>
+      <CustomToolNamespace>Microsoft.Xaml.XamlCompiler.CodeGenerators</CustomToolNamespace>
+      <LastGenOutput>WRL_XamlMetaDataProviderPass1.cs</LastGenOutput>
+    </None>
+    <None Include="Microsoft\Xaml\XamlCompiler\CodeGenerators\WRL_XamlMetaDataProviderPass2.tt">
+      <Generator>TextTemplatingFilePreprocessor</Generator>
+      <ClassNamespace>Microsoft.Xaml.XamlCompiler.CodeGenerators</ClassNamespace>
+      <CustomToolNamespace>Microsoft.Xaml.XamlCompiler.CodeGenerators</CustomToolNamespace>
+      <LastGenOutput>WRL_XamlMetaDataProviderPass2.cs</LastGenOutput>
+    </None>
+  </ItemGroup>
+  <ItemGroup>
+    <PackageReference Include="System.CodeDom" Version="9.0.4" />
+    <PackageReference Include="System.Reflection.Metadata" Version="9.0.4" />
+    <PackageReference Include="System.Runtime.InteropServices.RuntimeInformation" Version="4.3.0" />
+  </ItemGroup>
   <ItemGroup>
     <Reference Include="System.Xml" />
     <Reference Include="System.Core" />
@@ -20,4 +178,11 @@
     <Reference Include="Microsoft.Build.Framework" />
     <Reference Include="Microsoft.Build" />
   </ItemGroup>
+  <!-- Referencing a .NET Standard project includes a reference to System.Xaml, which conflicts with our private System.Xaml.  Remove the auto-included reference here -->
+  <Target Name="RemoveSystemXamlReference" BeforeTargets="ResolveAssemblyReferences">
+    <ItemGroup>
+      <Reference Remove="System.Xaml" />
+    </ItemGroup>
+  </Target>
+  <Import Project="$(MSBuildExtensionsPath)\Microsoft\VisualStudio\v17.0\TextTemplating\Microsoft.TextTemplating.targets" />
 </Project>
\ No newline at end of file
