diff --git a/Microsoft/UI/Xaml/Markup/Compiler/CodeGen/ArrayIndexStepCodeGenerator.cs b/Microsoft/UI/Xaml/Markup/Compiler/CodeGen/ArrayIndexStepCodeGenerator.cs
index f44b4da..13b49bd 100644
--- a/Microsoft/UI/Xaml/Markup/Compiler/CodeGen/ArrayIndexStepCodeGenerator.cs
+++ b/Microsoft/UI/Xaml/Markup/Compiler/CodeGen/ArrayIndexStepCodeGenerator.cs
@@ -1,5 +1,6 @@
 using Microsoft.Windows.UI.Xaml.Build.Tasks.Utilities;
 using Microsoft.Xaml.XamlCompiler;
+using Microsoft.Xaml.XamlCompiler.CodeGenerators;
 
 namespace Microsoft.UI.Xaml.Markup.Compiler.CodeGen;
 
@@ -9,16 +10,108 @@ internal class ArrayIndexStepCodeGenerator : BindPathStepCodeGenerator<ArrayInde
 	{
 		get
 		{
-			ICodeGenOutput parentPathExpression = base.Instance.Parent.CodeGen().PathExpression;
-			return new LanguageSpecificString(() => $"safe_cast<::Windows::Foundation::Collections::IVector<{base.Instance.ValueType.CppCXName()}>^>({parentPathExpression.CppCXName()})->GetAt({base.Instance.Index})", () => string.Format("{1}.as<::winrt::Windows::Foundation::Collections::IVector<{0}>>().GetAt({2})", base.Instance.ValueType.CppWinRTName(), parentPathExpression.CppWinRTName(), base.Instance.Index), () => $"{parentPathExpression.CSharpName()}[{base.Instance.Index}]", () => $"{parentPathExpression.VBName()}({base.Instance.Index})");
+			ICodeGenOutput parentPathExpression = Instance.Parent.CodeGen().PathExpression;
+			return new LanguageSpecificString(
+				() => $"safe_cast<::Windows::Foundation::Collections::IVector<{Instance.ValueType.CppCXName()}>^>({parentPathExpression.CppCXName()})->GetAt({Instance.Index})",
+				() => string.Format("{1}.as<::winrt::Windows::Foundation::Collections::IVector<{0}>>().GetAt({2})", Instance.ValueType.CppWinRTName(), parentPathExpression.CppWinRTName(), Instance.Index),
+				() => $"{parentPathExpression.CSharpName()}[{Instance.Index}]",
+				() => $"{parentPathExpression.VBName()}({Instance.Index})",
+				wrl: () =>
+				{
+					string n_temp = WRL_InternalGetTempVarName(nameof(PathExpression), "Temp"); // return value smart ptr
+					string n_vector = WRL_InternalGetTempVarName(nameof(PathExpression), "Vector");
+					return
+$$"""
+// BEGIN ArrayIndexStepCodeGenerator.PathExpression
+{{Instance.ValueType.WRLDeclareSmartVar(n_temp)}}
+if (SUCCEEDED(hrT))
+{
+{{parentPathExpression.WRLName().WRL_Indent(1)}}
+    if (SUCCEEDED(hrT))
+    {
+        ::Microsoft::WRL::ComPtr<::ABI::Windows::Foundation::Collections::IVector<{{Instance.ValueType.WRLNameTypeArg()}}>> {{n_vector}};
+        hrT = {{parentPathExpression.WRLOutVarName()}}->QueryInterface(IID_PPV_ARGS(&{{n_vector}}));
+        if (SUCCEEDED(hrT))
+        {
+            hrT = {{n_vector}}->GetAt({{Instance.Index}}, {{Instance.ValueType.WRLOutParam(n_temp)}});
+        }
+    }
+}
+{{Instance.ValueType.WRLNameNonSmartVar()}} {{PathExpression.WRLOutVarName()}} = {{Instance.ValueType.WRLGetSmartVar(n_temp)}};
+// END ArrayIndexStepCodeGenerator.PathExpression
+""";
+				},
+				wrlOutVar: () => WRL_InternalGetTempVarName(nameof(PathExpression), "Out"));
 		}
 	}
 
-	public override ICodeGenOutput UpdateCallParam => new LanguageSpecificString(() => string.Format("safe_cast<::Windows::Foundation::Collections::IVector<{0}>^>({1})->GetAt({2})", base.Instance.ValueType.CppCXName(), "obj", base.Instance.Index), () => string.Format("{0}.as <::winrt::Windows::Foundation::Collections::IVector<{1}>>().GetAt({2})", "obj", base.Instance.ValueType.CppWinRTName(), base.Instance.Index), () => string.Format("{0}[{1}]", "obj", base.Instance.Index), () => string.Format("{0}({1})", "obj", base.Instance.Index));
+	public override ICodeGenOutput UpdateCallParam
+	{
+		get
+		{
+			return new LanguageSpecificString(
+				() => string.Format("safe_cast<::Windows::Foundation::Collections::IVector<{0}>^>({1})->GetAt({2})", Instance.ValueType.CppCXName(), "obj", Instance.Index),
+				() => string.Format("{0}.as <::winrt::Windows::Foundation::Collections::IVector<{1}>>().GetAt({2})", "obj", Instance.ValueType.CppWinRTName(), Instance.Index),
+				() => string.Format("{0}[{1}]", "obj", Instance.Index),
+				() => string.Format("{0}({1})", "obj", Instance.Index),
+				wrl: () =>
+				{
+					string n_temp = WRL_InternalGetTempVarName(nameof(UpdateCallParam), "Temp"); // return value smart ptr
+					string n_vector = WRL_InternalGetTempVarName(nameof(UpdateCallParam), "Vector");
+					return
+$$"""
+// BEGIN ArrayIndexStepCodeGenerator.UpdateCallParam
+{{Instance.ValueType.WRLDeclareSmartVar(n_temp)}}
+if (SUCCEEDED(hrT))
+{
+    ::Microsoft::WRL::ComPtr<::ABI::Windows::Foundation::Collections::IVector<{{Instance.ValueType.WRLNameTypeArg()}}>> {{n_vector}};
+    hrT = obj->QueryInterface(IID_PPV_ARGS(&{{n_vector}}));
+    if (SUCCEEDED(hrT))
+    {
+        hrT = {{n_vector}}->GetAt({{Instance.Index}}, {{Instance.ValueType.WRLOutParam(n_temp)}});
+    }
+}
+{{Instance.ValueType.WRLNameNonSmartVar()}} {{UpdateCallParam.WRLOutVarName()}} = {{Instance.ValueType.WRLGetSmartVar(n_temp)}};
+// END ArrayIndexStepCodeGenerator.UpdateCallParam
+""";
+				},
+				wrlOutVar: () => WRL_InternalGetTempVarName(nameof(UpdateCallParam), "Out"));
+		}
+	}
 
 	public override ICodeGenOutput PathSetExpression(ICodeGenOutput value)
 	{
-		ICodeGenOutput parentPathExpression = base.Instance.Parent.CodeGen().PathExpression;
-		return new LanguageSpecificString(() => "safe_cast<::Windows::Foundation::Collections::IVector<" + base.Instance.ValueType.CppCXName() + ">^>" + $"({parentPathExpression.CppCXName()})->SetAt({base.Instance.Index}, {value.CppCXName()})", () => parentPathExpression.CppWinRTName() + ".as<::winrt::Windows::Foundation::Collections" + $"::IVector<{base.Instance.ValueType.CppWinRTName()}>>().SetAt({base.Instance.Index}, {value.CppWinRTName()})", () => PathExpression.CSharpName() + " = " + value.CSharpName(), () => PathExpression.VBName() + " = " + value.VBName());
+		ICodeGenOutput parentPathExpression = Instance.Parent.CodeGen().PathExpression;
+		return new LanguageSpecificString(
+			() => "safe_cast<::Windows::Foundation::Collections::IVector<" + Instance.ValueType.CppCXName() + ">^>" + $"({parentPathExpression.CppCXName()})->SetAt({Instance.Index}, {value.CppCXName()})",
+			() => parentPathExpression.CppWinRTName() + ".as<::winrt::Windows::Foundation::Collections" + $"::IVector<{Instance.ValueType.CppWinRTName()}>>().SetAt({Instance.Index}, {value.CppWinRTName()})",
+			() => PathExpression.CSharpName() + " = " + value.CSharpName(),
+			() => PathExpression.VBName() + " = " + value.VBName(),
+			wrl: () =>
+			{
+				string n_vector = WRL_InternalGetTempVarName(nameof(PathSetExpression), "Vector");
+				return
+$$"""
+// BEGIN ArrayIndexStepCodeGenerator.PathSetExpression
+if (SUCCEEDED(hrT))
+{
+{{(Instance.Parent.NeedsCheckForNull ? "/* Already null-checked */" : parentPathExpression.WRLName()).WRL_Indent(1)}}
+    if (SUCCEEDED(hrT))
+    {
+        ::Microsoft::WRL::ComPtr<::ABI::Windows::Foundation::Collections::IVector<{{Instance.ValueType.WRLNameTypeArg()}}>> {{n_vector}};
+        hrT = {{parentPathExpression.WRLOutVarName()}}->QueryInterface(IID_PPV_ARGS(&{{n_vector}}));
+        if (SUCCEEDED(hrT))
+        {
+{{value.WRLName().WRL_Indent(3)}}
+            if (SUCCEEDED(hrT))
+            {
+                hrT = {{n_vector}}->SetAt({{Instance.Index}}, {{value.WRLOutVarName()}});
+            }
+        }
+    }
+}
+// END ArrayIndexStepCodeGenerator.PathSetExpression
+""";
+			});
 	}
 }
